---
title: "A tour of DiffCoExpr"
author: "Vishali Umaiyalan"
date: "`r format(Sys.time(), '%d %b %Y')`"
output: 
  rmarkdown::html_vignette:
    toc: true
    number_sections: false
vignette: >
  %\VignetteIndexEntry{A tour of DiffCoExpr}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
library(knitr)
opts_chunk$set(fig.align = "center", 
               out.width = "90%",
               fig.width = 6, fig.height = 5.5,
               dev.args=list(pointsize=10),
               par = TRUE, # needed for setting hook 
               collapse = TRUE, # collapse input & ouput code in chunks
               warning = FALSE)
knit_hooks$set(par = function(before, options, envir)
  { if(before && options$fig.show != "none") 
       par(family = "sans", mar=c(4.1,4.1,1.1,1.1), mgp=c(3,1,0), tcl=-0.5)
})
set.seed(1) # for exact reproducibility
```
       
```{r setup}
require("devtools")
devtools::install_github("vishali-u/DiffCoExpr")
library(DiffCoExpr)

# Load packages needed for the tutorial
require(Seurat)
```

## Introduction

`DiffCoExpr` is an R package developed to construct co-expression networks from single cell RNA-seq data, and identify pairs of genes that are differentially co-expressed between pairs of co-expression networks. This can include constructing cell-type-specific co-expression networks and determining if genes that are co-expressed in both cell types are co-expressed at the same level or if the pairs are differentially co-expressed. The input to DiffCoExpr is gene-barcode matrices that are generated during alignment using the cellranger pipeline. By working with cellranger pipeline, `DiffCoExpr` will save users from having to perform many intermediate steps between alignment and co-expression construction. 

To download **DiffCoExpr**, use the following commands:

``` r
require("devtools")
devtools::install_github("vishali-u/DiffCoExpr", build_vignettes = TRUE)
library("DiffCoExpr")
```
To list all sample functions available in the package:
``` r
ls("package:DiffCoExpr")
```

To list all sample datasets available in the package:
``` r
data(package = "DiffCoExpr")
```

<br>

## Tutorial on 3k PBMCs from a Healthy Donor Datasest downloaded from 10X Genomics

This document will go over the steps needed to construct and compare co-expression networks generated from peripheral blood mononuclear cells (PBMCs) 
from a healthy donor. There are 2,700 cells in total in this dataset. 

### Step 1: Prepare the data
To construct a Seurat object containing normalized counts and annotated clusters using the sample data:
``` r
gene_bc_matrix_path <- system.file('extdata', 'pbmc', 'filtered_gene_bc_matrices', package = "DiffCoExpr")
cell_types_path <- system.file('extdata', 'pbmc', 'cell_types.csv', package = "DiffCoExpr")
srat <- prepareData(gene_bc_matrix_path = gene_bc_matrix_path, cell_types_path = cell_types_path)
```
To examine the cell identities in the Seurat object:
``` r
Idents(srat)
```

### Step 2: Construct the expression and correlation matrices
We can limit our data to just one cell type. Here we are only interested in constructing the matrices for B cells. To create expression and correlation matrices:
``` r
expr_matrix <- get_expression_matrix(srat = srat, cell_type = "B")
corr_matrix <- get_correlation_matrix(expression_matrix = expr_matrix)
```

To visualize the expression and correlation across cells and gene pairs, respectively:
``` r
expr_plot <- plotExpressionMatrix(expression_matrix = expr_matrix)
corr_plot <- plotCorrelationMatrix(correlation_matrix = corr_matrix)
```

### Step 3: Construct a co-expression network 
To create a list of co-expressed genes and the correlation between the genes:
``` r
edge_list <- buildCoexpressionNetwork(correlation_matrix = corr_matrix)
```

### Step 4: Create a co-expression network graph
TODO: show construction on a full network and not just a subset once the plotCoexpressionNetwork is working well with large datasets
To create a co-expression network graph on a subset of our data:
```r
subset_edges <- edge_list[1:30,]
network_plot <- plotCoexpressionNetwork(subset_edges)
```

## Package References

- Umaiyalan, V. (2023) DiffCoExpr: Construct and compare co-expression networks
  from scRNA-seq. Unpublished. https://github.com/vishali-u/DiffCoExpr

<br>

## Other References

- Butler A, Hoffman P, Smibert P, Papalexi E, Satija R (2018) Integrating single-cell transcriptomic data across different conditions, technologies, and species. *Nature Biotechnology* 36, 411-420. https://doi.org/10.1038/nbt.4096

- Butler, A. (2015). Seurat: Tools for Single Cell Genomics R package. https://cran.r-project.org/web/packages/Seurat

- Chowdhury H., Bhattacharyya D., Kalita J (2020) (Differential) Co-Expression Analysis of Gene Expression: A survey of Best Practices. *IEEE/ACM Transactions on Computational Biology and Bioinformatics* 17. https://ieeexplore.ieee.org/document/8613814

- Contreras-Lopez, Moyano T., Soto D., Gutierrez R (2018) Step-by-Step Construction of Gene Co-expression Networks from High-Throughput Arabidopsis RNA Sequencing Data. *Methods in Molecular Biology* https://pubmed.ncbi.nlm.nih.gov/29525965/

- Eisen M., Spellman P., Brown P., Botstein D (1998) Cluster analysis and display of genome-wide expression patterns. *Biological Sciences* 95. https://www.pnas.org/doi/10.1073/pnas.95.25.14863

- Hennetin J., Pehkonen P., Bellis M (2009) Construction and use of gene expression covariation matrix. *BMC Bioinformatics* 10. https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-10-214

- Lemoine G., Scott-Boyer M., Ambroise B., Perin O., Droit A (2021) GWENA: gene co-expression networks analysis and extended modules characterization in a single Bioconductor package. *BMC Bioinformatics* 22. https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-021-04179-4

- R Core Team (2023). R: A language and environment for statistical computing. R Foundation for Statistical Computing, Vienna, Austria. https://www.R-project.org/

- Su C., Xu Z., Shan X., Cai B., Zhao H., Zhang J (2023) Cell-type-specific co-expression inference from single cell RNA-sequencing data. *Nature Communications* 14. https://www.nature.com/articles/s41467-023-40503-7

- Wickham, H. and Bryan, J. (2019). *R Packages* (2nd edition). Newton, Massachusetts: O'Reilly Media. https://r-pkgs.org/

----

```{r}
sessionInfo()
```